### PR TITLE
Support OpenAI's new O1 Model Series

### PR DESCRIPTION
**Pull request by debanjum**
## Major
- The new O1 series doesn't seem to support streaming, response_format enforcement, 
  stop words or temperature currently. 

- Remove any markdown json codeblock in chat actors expecting json responses

## Minor
- Override block display styling of links by Katex in chat messages

### CODE REVIEW

File: src/khoj/processor/conversation/openai/utils.py [hunk 1]
```diff
@@ -45,24 +45,35 @@ def completion_with_backoff(
...
+        for chunk in chat:
+            if len(chunk.choices) == 0:
+                continue
+            delta_chunk = chunk.choices[0].delta  # type: ignore
+            if isinstance(delta_chunk, str):
+                aggregated_response += delta_chunk
+            elif delta_chunk.content:
+                aggregated_response += delta_chunk.content
 
     return aggregated_response
```
Comment (sabaimran) [inline src/khoj/processor/conversation/openai/utils.py:78]:
```suggestion
    if not stream:
        return chat.choices[0].message.content
    for chunk in chat:
        if len(chunk.choices) == 0:
            continue
        delta_chunk = chunk.choices[0].delta  # type: ignore
        if isinstance(delta_chunk, str):
            aggregated_response += delta_chunk
        elif delta_chunk.content:
            aggregated_response += delta_chunk.content
    return aggregated_response
```

nit: I find early return a little bit more readable

File: src/khoj/processor/conversation/openai/utils.py [hunk 2]
```diff
@@ -112,24 +123,35 @@ def llm_thread(g, messages, model_name, temperature, openai_api_key=None, api_ba
...
 
         formatted_messages = [{"role": message.role, "content": message.content} for message in messages]
+        stream = True
+
+        # Update request parameters for compatability with o1 model series
+        if model_name.startswith("o1"):
+            stream = False
+            temperature = 1
+            model_kwargs.pop("stop", None)
+            model_kwargs.pop("response_format", None)
```
Comment (sabaimran) [inline src/khoj/processor/conversation/openai/utils.py:133]: Refer to documentation on these parameter decisions here. Let's keep an eye on it as the beta rolls out: https://platform.openai.com/docs/guides/reasoning/beta-limitations

Comment (sabaimran) [reply src/khoj/processor/conversation/openai/utils.py:133]: Maybe add a link to the doc in the code? Will make it easier to remove the custom code later on hopefully.

File: src/khoj/routers/helpers.py [hunk 3]
```diff
@@ -287,8 +287,8 @@ async def aget_relevant_information_sources(
 
     try:
         response = response.strip()
-        # Remove any markdown json codeblock formatting if present (useful for gemma-2)
-        if response.startswith("```json"):
+        # Remove any markdown json codeblock formatting if present. Useful for non schema enforceable models
+        if response.startswith("```json") and response.endswith("```"):
```
Comment (sabaimran) [inline src/khoj/routers/helpers.py:291]: Future optimization, maybe we can move this custom logic to a helper method for code re-use/maintainability.

**Review by sabaimran — APPROVED** [review]
Couple suggestions, but looks good. Excited to get this out.

File: src/khoj/routers/helpers.py (hunk 3 above)
Comment (debanjum) [reply src/khoj/routers/helpers.py]: Good point. Put that json codeblock removal code snippet under conversation utils as a helper function for reusability

**Pull request merged**